tokenizer_ts/src/bytePairEncode.ts

```diff
@@ -77,14 +79,25 @@ export function bytePairEncode(
     return [ranks.get(mergingBytes)!];
   }
 
+  let minRank = maxRank;
+  let minIndex = -1;
+
   const byteIndicesAndRanks: [number, number][] = [];
-  for (let i = 0; i < mergingBytes.length + 1; i++) {
-    byteIndicesAndRanks.push([i, Number.MAX_SAFE_INTEGER]);
+  for (let i = 0; i < mergingBytes.length - 1; i++) {
+    const rank = ranks.get(mergingBytes.subarray(i, i + 2)) ?? maxRank;
+    if (rank < minRank) {
+      minRank = rank;
+      minIndex = i;
+    }
+
+    byteIndicesAndRanks.push([i, rank]);
   }
+  byteIndicesAndRanks.push([mergingBytes.length - 1, maxRank]);
+  byteIndicesAndRanks.push([mergingBytes.length, maxRank]);
 
   function getRank(startIndex: number, skip = 0): number {
     if (startIndex + skip + 2 < byteIndicesAndRanks.length) {
-      const slice = mergingBytes.slice(
+      const slice = mergingBytes.subarray(
```

Comment:
This change was the bulk of the performance benefit in this PR: slice copies the array, whereas subarray just references the underlying memory. I looked elsewhere and didn't see other bad usages of `slice()` in this code path.

Reply:
Interesting- I didn't even know about subarray :O